benchmarks/distributed/test_distributed.py

```diff
@@ -0,0 +1,130 @@
+
+def test_max_actors():
+    @ray.remote
+    class Actor:
+        def foo(self):
+            pass
+
+    print("Launching actors")
+    actors = [Actor.remote() for _ in trange(MAX_ACTORS_IN_CLUSTER)]
+
+    print("Ensuring actors have started")
+    for actor in tqdm(actors):
+        assert ray.get(actor.foo.remote()) is None
+
+def test_max_running_tasks():
+    counter = Semaphore.remote(0)
+    blocker = Semaphore.remote(0)
+    total_resource = ray.cluster_resources()["node"]
+
+    # Use this custom node resource instead of cpus for scheduling to properly loadbalance. CPU borrowing could lead to unintuitive load balancing if we used num_cpus.
+    resource_per_task = total_resource / MAX_RUNNING_TASKS_IN_CLUSTER
+
+    @ray.remote(num_cpus=0, resources={"node": resource_per_task})
+    def task(counter, blocker):
+        sleep(30)
+        # TODO (Alex): We should guarantee that all these tasks are running concurrently. Unfortunately these semaphores currently don't work due to too many simultaneous connections.
+        # ray.get(counter.release.remote())
+        # ray.get(blocker.acquire.remote())
+        pass
+
+    print("Launching tasks")
+    refs = [task.remote(counter, blocker) for _ in trange(MAX_RUNNING_TASKS_IN_CLUSTER)]
+
+    # print("Ensuring all tasks are running")
+    # for _ in trange(MAX_RUNNING_TASKS_IN_CLUSTER):
+    #     ray.get(counter.acquire.remote())
+
+    # print("Unblocking all tasks")
+    # for _ in trange(MAX_RUNNING_TASKS_IN_CLUSTER):
+    #     ray.get(blocker.release.remote())
+    print("Waiting ...")
+    for _ in trange(30):
+        sleep(1)
+
+    print("Ensuring all tasks have finished")
+    for _ in trange(MAX_RUNNING_TASKS_IN_CLUSTER):
+        done, refs = ray.wait(refs)
+        assert ray.get(done[0]) is None
+
+def test_many_placement_groups():
+    print("Creating placement groups")
+    placement_groups = []
+    for _ in trange(MAX_PLACEMENT_GROUPS):
+        bundle1 = {"CPU": 0.5, "node": 0.125}
+        bundle2 = {"CPU": 0.5}
+        bundle3 = {"node": 0.125}
+        pg = placement_group([bundle1, bundle2, bundle3], strategy="STRICT_SPREAD")
+        placement_groups.append(pg)
+
+    print("Waiting for all placement groups to be ready")
+    for pg in tqdm(placement_groups):
+        ray.get(pg.ready())
+
+
+    @ray.remote(num_cpus=0.5, resources={"node": 0.125})
+    def f1():
+        sleep(30)
+
+    @ray.remote(num_cpus=0.5)
+    def f2():
+        sleep(30)
+
+    @ray.remote(resources={"node": 0.125})
+    def f3():
+        sleep(30)
+
+    print("Running tasks on placement groups")
+    refs = []
+    for pg in tqdm(placement_groups):
+        ref1 = f1.options(placement_group=pg).remote()
+        ref2 = f2.options(placement_group=pg).remote()
+        ref3 = f3.options(placement_group=pg).remote()
+        refs.extend([ref1, ref2, ref3])
+
+    print("Waiting...")
+    for _ in trange(30):
+        sleep(1)
+
+    print("Getting results")
```

Comment:
Btw, you need to explicitly clean up placement groups (if you share tests to a single cluster) because they are removed only when the job is terminated.

Reply:
good thing it's the last test :p 

but thanks, i'll make sure to do this